### PR TITLE
Animated WebP support

### PR DESCRIPTION
**Pull request by garrettmoon**

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ViewController.m:</p>
<pre><code>//
//  ViewController.m
//  Texture
//
//  Copyright (c) 2014-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under the BSD-style license found in the
//  LICENSE file in the /ASDK-Licenses directory of this source tree. An additional
//  grant of patent rights can be found in the PATENTS file in the same directory.
//
//  Modifications to this file made after 4/13/2017 are: Copyright (c) 2017-present,
//  Pinterest, Inc.  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/222221111587c457d9f2257314c06daf/raw/68089460caeee2c68e3de6b7add02205ea347a03/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ViewController.m:</p>
<pre><code>//
//  ViewController.m
//  Texture
//
//  Copyright (c) 2014-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under the BSD-style license found in the
//  LICENSE file in the /ASDK-Licenses directory of this source tree. An additional
//  grant of patent rights can be found in the PATENTS file in the same directory.
//
//  Modifications to this file made after 4/13/2017 are: Copyright (c) 2017-present,
//  Pinterest, Inc.  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by djblake** [comment]
@garrettmoon from my testing:
### Fixed:
- Transition behaviour now working properly again, button images are no longer being unloaded when a button is removed from the layer hierarchy.
- No longer seeing flashing when removing an image from one parent and directly adding it to another parent. 

### Possible issues:
- When moving a gif from one parent and immediately adding it to another I still see a flash (a double flash actually) as the image is presumably being unloaded when `removeFromSupernode:` triggers the `didExitDisplayState` which removes the cover image. Not sure if anything can be done about this as obviously still need the memory benefit of doing this when moving `ASCellNode`s outside of display range. Unfortunately `displaysAsynchronously = YES` seems to have no effect on the flashing like it did when the regular images had this problem. That would probably be the optimum fix for this if it worked. A hack for this is probably if I place the the gifs in a container `ASDisplayNode` and call `removeFromSupernode:` on that node instead, fooling the gifs into thinking they still belong to the same parent and never calling `didExitDisplayState`.

- Not sure if anything can be done about this, but I can still crash (due to memory overload) that test example project I made with the 1000 gifs in a `ASCollectionNode` if I scroll as fast as I can to the bottom. Only happens when loading the gif through `.URL =`, presumably because the loading from the memory cache occurs after its already left the display range. Not sure how much of a real world problem this would actually be, but something to note as a potential pitfall anyway. 


Other than those, all looking good!

**Comment by garrettmoon** [comment]
Thanks for the feedback @djblake! The second issue you brought up is something I want to look at. The first is a little harder to deal with. I could make displaysAsynchronously NO work, but it probably would really mess up the performance of any transitions. I'd recommend a workaround of copying the node's layer contents before removing it and setting them after adding it back in.

**Comment by garrettmoon** [comment]
@djblake Turns out #2 is actually a separate issue, probably only encountered more often on GIFs! Fix is here: https://github.com/TextureGroup/Texture/pull/606

File: examples/AnimatedGIF/ASAnimatedImage/ViewController.m [hunk 1]
```diff
@@ -32,7 +30,9 @@ - (void)viewDidLoad {
   // Do any additional setup after loading the view, typically from a nib.
   
   ASNetworkImageNode *imageNode = [[ASNetworkImageNode alloc] init];
-  imageNode.URL = [NSURL URLWithString:@"https://s-media-cache-ak0.pinimg.com/originals/07/44/38/074438e7c75034df2dcf37ba1057803e.gif"];
+//    imageNode.URL = [NSURL URLWithString:@"https://s-media-cache-ak0.pinimg.com/originals/07/44/38/074438e7c75034df2dcf37ba1057803e.gif"];
+//  imageNode.URL = [NSURL URLWithString:@"https://storage.googleapis.com/downloads.webmproject.org/webp/images/dancing_banana2.lossless.webp"];
```
Comment (nguyenhuy) [inline examples/AnimatedGIF/ASAnimatedImage/ViewController.m]: Let's remove these lines.

Comment (garrettmoon) [reply examples/AnimatedGIF/ASAnimatedImage/ViewController.m]: I think I'll remove one and add a comment to the other one saying "uncomment to see animated webp support"?

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/a1997ca35d73cffad17bf051d53dd63f/raw/679f77e527cc0f0ae7fbb48f91f76d33c4b76dd0/log.txt)

**Pull request merged**